core/src/main/scala-2/zio/config/TupleConversion.scala

```diff
@@ -0,0 +1,86 @@
+package zio.config
+
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
+
+trait TupleConversion[A, B] {
+  def to(a: A): B
+  def from(b: B): A
+}
+object TupleConversion {
+  def apply[Prod, Tup]: TupleConversion[Prod, Tup] = macro genTupleConversion[Prod, Tup]
+
+  def genTupleConversion[Prod: c.WeakTypeTag, Tup: c.WeakTypeTag](
+    c: whitebox.Context
+  ): c.Expr[TupleConversion[Prod, Tup]] = {
+    import c.universe._
+    val prodTpe = c.weakTypeOf[Prod]
+    if (!prodTpe.typeSymbol.isClass ||
+        !prodTpe.typeSymbol.asClass.isCaseClass) {
+      c.abort(c.enclosingPosition, s"Type ${prodTpe.typeSymbol} is not a case class")
+    }
+    val paramLists = prodTpe.typeSymbol.asClass.primaryConstructor.asMethod.typeSignatureIn(prodTpe).paramLists
+    val result = paramLists match {
+      case List(List(singleParam)) =>
+        // Special case: single parameter products
+
+        val typ = singleParam.typeSignatureIn(prodTpe).finalResultType
+
+        q"""new _root_.zio.config.TupleConversion[$prodTpe, $typ] {
+                override def to(a: $prodTpe): $typ = a.${singleParam.name.toTermName}                
+                override def from(b: $typ): $prodTpe =
+                    new ${prodTpe}(b)
+        }
+        """
+
+      case List(params) =>
+        // Generic case: n > 1 parameters
+
+        val tupleName = definitions.TupleClass(params.size).name.toTypeName
+        val tupleParams = params.map { sym =>
+          sym.typeSignatureIn(prodTpe).finalResultType
+        }
+        val tup = tq"$tupleName[..$tupleParams]"
+        val packers =
+          params.map { sym =>
+            val symTerm = sym.name.toTermName
+            q"a.$symTerm"
+          }
+        val unpackers =
+          params.indices.map { idx =>
+            val accessor = TermName(s"_${idx + 1}")
+            q"b.$accessor"
+          }
+
+        q"""new _root_.zio.config.TupleConversion[$prodTpe, $tup] {
+                override def to(a: $prodTpe): $tup =
+                (..$packers)
+                override def from(b: $tup): $prodTpe =
+                new ${prodTpe}(..$unpackers)
```

Comment:
Should we need a `new` here?

Reply:
The initial version called the companion object's apply instead but it did not work in all cases. In some tests the companion object was `null`, maybe because of a similar issue than the one found in Scala 3, with the macro being invoked from base constructor invocation. This way it is safe and always works.

Reply:
Sure. Makes sense.